### DISCUSSION TITLE
Consider alternative versioning schemes?

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#11280] [General]
Posting this discussion to take the temperature for looking in to adopting a new versioning scheme. 

Some comments and issues regarding the latest release have come up, and also this article by Hynek Schlawack was posted today: 
 
https://hynek.me/articles/semver-will-not-save-you/

I really can't argue with much in that article, and in particular note that for Bokeh we are definitely making judgment calls about things from time to time (or more directly: rationalizing why changes don't *really* merit a major version bump). If I am honest, I would rather stop pretending that we do (or can) adhere to strict semver. 

The article links to this review of different calver adoptions:

https://calver.org/

Since we have CDN coordination to consider, perhaps we could/should consider a scheme that is relatively consistent with our current machinery, e.g 

* `YYYY.MM.patch`
* `YY.minor.patch`

Thoughts @bokeh/dev?

**Comment by p-himik** [comment]
Thanks for the article. It has some valid points and I agree with its stance on SemVer. So much so that when updating dependencies of my own projects, I not only read changelogs but sometimes even diffs. After all, if a version change is a tldr for a changelog, a changelog is a tldr for a series of commits.

> I would rather stop pretending that we do (or can) adhere to strict semver

If only more developers were as brave as you to admit this. :smile: 

CalVer does sound nice. But what do you mean by `patch`? Just an ever increasing number?
In Clojure world, some libraries just make that `patch` be the number of commits before that version. I think such an approach is a tiny bit better because it somewhat helps to assess the amount of changes between two versions.

**Comment by bryevdv** [comment]
> CalVer does sound nice. But what do you mean by patch? Just an ever increasing number?

That's all I meant by it, e.g if we aim for ~4 "big" releases a year, then a "patch" is just a small release in between. But I am open to any/all other ideas here. We could have it mean "commits since last `YY.mm`" which is also increasing.

> If only more developers were as brave as you to admit this. 

FWIW I think there are classes of projects where it can be successful, those that: are very small, have few/no dependencies, are basically just functions w/o side effects. But Bokeh is not that. :) To really, truly, actually try to make promises we'd have to test for:

> (dozens of browser versions) * (many OS versions) *  (4-5 Python versions) *  (several Tornado versions) * (lots of notebook versions + lots of jlab versions) 

And then for that huge matrix have tens of thousands of unit tests, thousands of example tests, hundreds of integration tests, dozens of entire downstream project test suites, untold image diff tests... the resources required are just in a different galaxy.

**Comment by p-himik** [comment]
> We could have it mean "commits since last `YY.mm`" which is also increasing.

Sounds good to me. A "commits in total" would also work if you're OK with potentially large numbers. Just as an example of such numbers: https://github.com/clojure/tools.deps.alpha/releases

**Comment by bryevdv** [comment]
I'm not personally a huge fan of big numbers in versions, especially since our numbers would be even larger, also we may want to release off previous branches occasionally. So I'd lean personally lean towards "meaninless incrementing integer" or "commits since last 'big' version".

**Comment by bryevdv** [comment]
I'd also like to make it clear that I *do think* the aspiration of maintaining compatibility should be a "north star" for our work. I don't think we should make breaking changes casually. I do think we should strive to be intentional, with deprecation periods and communicating detailed information about changes. But occasionally more sudden changes are warranted, and often changes slip in w/o consideration. So this is really just about recognizing that reality.

**Comment by p-himik** [comment]
Absolutely. Sometimes people even go as far as avoiding incompatible changes altogether by introducing new packages (similar how it's happening in Go, according to that article). It might be something to consider as well - having `bokeh3 v0` instead of `bokeh v3.0`.

**Comment by bryevdv** [comment]
Also for reference,  here is the discussion on the Dask tracker: https://github.com/dask/community/issues/100

**Comment by mattpap** [comment]
To me calendar-based versioning schemes seem awkward, if the release schedule is not bound to the calendar. So, Ubuntu's YY.MM (e.g. 20.04) is fine, but otherwise both `YYYY.MM.patch` and `YY.minor.patch` don't convey meaning if you just have a single version number. E.g. 21.1.5, is this a date, or just year/month and some other arbitrary integer? In the case of `YYYY.MM.patch`, what if we need multiple non-patch releases per month?

`<draft>`
I think this is just a complication of a much simpler and much more fundamental versioning scheme. Thus I would like propose `major.0.patch` scheme. Ideally it would be just `major.patch`, but `0` is needed for compatibility with semver, especially npm. In this scheme every release is a major, potentially breaking, release, followed by a sequence of non-breaking patch releases. Major releases are the default and patch releases are there just to amend on major releases (strictly no new features, no API changes, etc.). If one is worried that `major` would grow too quickly, then with the current cadence we would be at 23 (maybe a little higher due to 0.x versioning oddity). A major assumption is that `major` is a continuous sequence of integers and users always install the latest `patch` for a given `major` release. So, e.g. if there's a botched `N.0.0` release then we follow up with `N.0.1`.

In this scheme, every feature/API/etc. has a life span assigned as an integer interval `[major_add(f), major_deprecate(f)]`. There's an implicit `major_remove(f)` (and an associated implicit interval), which is defined as `major_deprecate(f) + k`, where `k` is an integer constant derived from the typical release cadence of a project. Thus if we release very often then `k` is higher, if not so often, then `k` is smaller. For a given set of required features `F = {f_0, ..., f_n}` by a downstream project, users, etc., one can compute a intersection of their life span intervals to determine which versions of the project are supported.

So much for an ad-hoc, idealistic definition. The point is that we admit that we break things in minor releases (for various reasons) all the time and that won't change. The versioning scheme has to admit that as well and give our users the ability to predict what works and when.
`</draft>`

**Comment by p-himik** [comment]
> E.g. 21.1.5, is this a date, or just year/month and some other arbitrary integer?

That's the point - it doesn't matter at all. There's a need to have a version in the format of `x.x.x` and there's a need to make it ordered. That's it. The word "calver" would also be useful to immediately convey to users that the version means absolutely nothing and that they must look at the changelog to fugure out their next steps when upgrading to any version.

> In the case of YYYY.MM.patch, what if we need multiple non-patch releases per month?

This question made me realize that Bokeh supports older versions and backports/creates some fixes.
Hmm. Is that practice a very important one to keep?
After switching to a new version format, no matter which, how many old releases will have to be supported? Because if previously you could say something like "we will publish only important fixes for Bokeh 2 when Bokeh 3 is the current version", now saying something like that but only for Bokeh 27 and 28 (or 2021.03 and 2021.04) won't mean much because Bokeh 29 (or 2021.05) will be released in a month. It seems to me that it would be better to drop any old version support and just fix any issues in the newest release.

**Comment by bryevdv** [comment]
> This question made me realize that Bokeh supports older versions and backports/creates some fixes.
Hmm. Is that practice a very important one to keep?

Yes and no. This (new-ish) practice is not so much about supporting old branches, as it is about affording some additional flexibility to release more frequently. We have now gotten in to a situation *many, many times* where our "next" release branch has gotten much bigger than planned, and gets stuck in a state where it cannot be released yet. When this happens we are hamstrung, and this has prevented us from getting important fixes that were already merged out to users in timely fashion. 

There are really two cadences we need to accommodate:

* fast loop: small (but important) bug or docs fixes, critical bug fixes of any size
* slow loop: larger feature work, infrastructural work or refactoring, risky work requiring more testing

As a side effect, pushing bug fixes and small updates into point releases on the last release branch allows the "distance" between the last release release branch and the upcoming release branch to be smaller than it otherwise would be (and thus less risky).

But the main point to note is that we really only envision issuing those "point" releases on the previous branch up until the next release branch is released. Then point releases on the old branch are done, so i.e it's not really about "supporting" old branches. 

So as for @mattpap it needs some elaboration with the above in mind. How do we accommodate both long and short term dev loops? Do you propose something like landing all dev on a dedicated perpetual "dev" branch, and periodically cherry picking commits back to `main` to tag releases there?

**Comment by bryevdv** [comment]
I guess I should mention, the default option, which is basically just to keep the status quo, but explain clearly that we **do not** adhere to strict semver. Then I would decribe `x.y.z` in terms:

* Z is for fast-cycle bugfixes and docs only
* Y is for larger, slower-cycle work. We strive for compatibility, but do not guarantee it. We will exercise discretion on intentional changes, prioritize changes that affect the smallest number of users,  and communicate them clearly
* X large breaking changes affecting all users to the library for sure

**Comment by p-himik** [comment]
I would switch to the option offered by mattpap then. Dependency resolution tools impose their own semantics, even if the users don't want it. Changing the major version much more often would decrease any possible effects of those semantics.

**Comment by bryevdv** [comment]
OK here is another proposal. It's similar to @mattpap except intead of perpetual Y=0, it's a perpetual Z=0

So releases are `6.0.0`, `6.1.0`, `6.2.0`, `7.0.0`, `8.0.0`, `8.1.0`, `8.2.0`, `8.3.0`, ...

This gives us:

* X = slow-loop feature development
* Y = fast-loop docs, small, isolated, or critical bugfixes
* Z = always zero

The reason I would prefer this is that bumping Z instead Y makes the version more tedious when having to talk about them with human beings. I'd literally rather people be able to say "ten point two" not "ten point oh point two" 

Basically this just would shift the current understanding of release levels "up one" i.e. in practice we have been using minor releases as major release, so this would just acknowledge that. I *do* think we can make *small* fast-loop minor releases with some frequency that *do* have an acceptably very low risk of breakage. Right now those are point releases but this would make them minor releases.

**Comment by bryevdv** [comment]
cc: would really love to hear from holoviz folks (@philippjfr @jbednar etc)  and dask folks (@quasiben @jsignell etc)

**Comment by jsignell** [comment]
Dask recently changed to CalVer in December and so far so good. I don't think we've had any complaints and it aligns much more cleanly with what was already happening in practice.

This is the format the dask landed on:

> `YYYY.MM.X`, where `X` is an incrementing identifier for releases in the same month, would allow us to release as often as we'd like (e.g. multiple releases on the same day) and is PEP 440 compliant....we 0-index our incrementing identifier

https://github.com/dask/community/issues/100#issuecomment-704531301

This format has the nice features of being a very obvious date and being string sortable.

**Comment by jsignell** [comment]
There is a separate discussion in dask around the concept of "blessed releases". These would be releases that are far from major changes and therefore less buggy. I think we arrived at noting these releases after the fact in a document rather than trying to fit them into the version scheme. This makes sense for dask because we often don't know ahead of time which releases will be pretty good. It takes a week or two for that to become fully apparent. I'll just note that this is 100% a judgement call, and might not make sense at all for bokeh. 

ref: https://github.com/dask/community/issues/101

**Comment by bryevdv** [comment]
Thanks for your comments @jsignell I think a blessed release idea is an interesting one and may definitely make sense if we move away from "semver". Agree that's definitely a post-hoc judgment for the team to make.

**Comment by jbednar** [comment]
Speaking from a HoloViz perspective, we've now started pinning to the Bokeh minor release, i.e. the latest Panel (and thus HoloViews and hvPlot) releases pin to Bokeh <2.4.  Bryan's modification of Mateusz's proposal, i.e. fixing the patch level at 0, makes sense, but we'd then need to pin <3 instead. Either way, we can deal with it.

If Bokeh does follow Bryan's suggestion, then non-zero patch levels remain open for the most critical and most minor bugfixes/patches in the rare cases when those are needed. I.e., typically we'd expect to see .0 at the end, acknowledging that any non-trivial release is also a potentially breaking release, but if we had some tiny bugfix to patch onto an existing release then x.y.1, x.y.2, etc. remain available. 

Seems like that way we'd respect SemVer, acknowledge that it's rare to have a truly non-breaking release, and yet not just give up on everything as CalVer does. E.g. I don't know what we'd do in Panel if Bokeh does CalVer; just pin to the specific version available on the date Panel is released, and accept no future versions?  Bryan or Mateusz's proposals seem preferable to that, as it gives us at least _some_ chance of releasing bugfixed Bokeh without having to re-release Panel.

**Comment by bryevdv** [comment]
So I've had a chance to let this marinate on this a bit, and have a few further thoughts. 

I am feeling less optimistic about a sweeping change like calver, partially because I think it would fall to me to drive that effort and I am just not up to it, given the need to coordinate external tools and resources like BokehJS CDN. There is just alot of technical risk to changing anything build related and it is not worth the stress. 

However, the above proposals also do not sit perfectly well with me either, because I do think what the "status quo" offers is three distinct levels of signaing:

X = huge, very big, definitely breaking, especially visuals, strong need to telegraph "big event"
Y = regular feature development pace, may break a few things, *but usually mostly* for projects, not users
Z = rapid schedule, small, much less risk 

I think having those three levels available to us, *regardless of semver*, is actually very good. Conversely: I think both @mattpap and my proposals above destroy any possiblility of over having "major major" releases, and I think that is bad. 

Regarding the "Y" level, some points: 

* I do 100% think, especially at this point, that most changes affect downstream projects and not day to day users. There is literally common use-case code from 6 years and two major release cycles ago that runs today without any chagnes:

  https://docs.bokeh.org/en/0.8.2/docs/gallery/image.html
  https://docs.bokeh.org/en/0.8.2/docs/gallery/iris.html
  https://docs.bokeh.org/en/0.8.2/docs/gallery/scatter.html
  https://docs.bokeh.org/en/0.9.3/docs/gallery/texas.html

  The places where normal users have gotten bitten is with things that probably should have been called "experimental" in the first places (e.g. custom extensions). We could/shoud stop shying away from labeling experimental things as such. 

* Alternatively downstream projects with deep integrations face higher risk. But no change here will change that @jbednar acknowledges they take this in to account already:
  > Speaking from a HoloViz perspective, we've now started pinning to the Bokeh minor release,

I think maybe an option here is just to continue being judicious, and conservative with respect to normal users, and to also pro-actively work with downstream projects to help keep things coordinated. Which is advised anyway.

I guess my position is that we should just adopt the process that *works for us regardless of any external considerations*, and simply explain that process as clearly and succinctly as possible. I've been trying to think of what simple organizing principle might encompass this, and a discussion here https://github.com/bokeh/bokeh/issues/11152#issuecomment-815957091 did give me a new idea. 

* X : There *will* be breaking changes. And especially: *visual* breaking changes
* Y : There probably won't be breaking changes. If there are, they will probably not affect normal users, but *will not* involve visual changes.
* Z : Very small, low risk bugfixes on a faster schedule

What occurred to me was specifically this: I think the "minor bump means no visual changes" is actually *exactly the guarantee are best set up to provide* with all of our expanded and expanding visual tests. I think we should just not fight against that reality, and embrace and leverage it as our easily explainable criteria.

Edit: and in case it was not evident: **I think we should be 100% clear, up-front,  and transparent that we do not adhere to semver.**

**Comment by bryevdv** [comment]
TLDR; I am circled-back at a defense of the status quo, but with a more clearly delineated rationale:
* maintain capability to signal three levels of development *appropriate to our project*

And also with more clearly delineated criteria:
* major bump specifically and definitely means *visual changes* (e.g. a switch to managed/unmanaged layout)
  * which is also what we can maintain with visual tests

**Comment by bryevdv** [comment]
cc @mattpap see my last two comments from last month
